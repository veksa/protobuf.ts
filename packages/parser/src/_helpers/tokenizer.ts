enum Block {
    none,
    comment,
    commentMulti,
    oneString,
    twoString,
}

const EMPTY: Record<string, boolean> = {
    ' ': true,
    '\t': true,
    '\n': true,
};

export const SYMBOLS: Record<string, boolean> = {
    '=': true,
    ',': true,
    ':': true,
    ';': true,
    '[': true,
    ']': true,
    '(': true,
    ')': true,
    '{': true,
    '}': true,
    '<': true,
    '>': true,
};

// eslint-disable-next-line complexity
export function tokenizer(source: string) {
    const tokens: string[] = [];
    const lines: number[] = [];
    const columns: number[] = [];
    let tmp = '';
    let block: Block = Block.none;
    let sameLine = true;
    let curLine = 1;
    let column = 1;

    const finalize = (after = false, force = false) => {
        if (tmp.length > 0 || force) {
            tokens.push(tmp);
            lines.push(curLine);
            columns.push(after ? column : column - tmp.length);
            tmp = '';
        }
    };

    for (let i = 0; i < source.length; i++) {
        const cur = source[i];
        const next = source[i + 1];

        switch (true) {
            case block === Block.none && cur + next === '//':
                finalize();
                block = Block.comment;
                tmp += sameLine ? '!//' : '//';
                finalize(true);
                i++;
                column++;
                break;

            case block === Block.comment && cur === '\n':
                block = Block.none;
                finalize(false, true);
                break;

            case block === Block.none && cur + next === '/*':
                block = Block.commentMulti;
                finalize();
                tmp = '/*';
                finalize(true);
                i++;
                column++;
                break;

            case block === Block.commentMulti && cur + next === '*/':
                block = Block.none;
                finalize(false, true);
                tmp = '*/';
                finalize(true);
                i++;
                column++;
                break;

            case block === Block.none && cur === '\'':
                finalize();
                tmp = cur;
                block = Block.oneString;
                break;

            case block === Block.oneString && cur + next === '\\\'':
                tmp += '\'';
                i++;
                break;

            case block === Block.oneString && cur === '\'':
                tmp += cur;
                finalize();
                block = Block.none;
                break;

            case block === Block.none && cur === '"':
                finalize();
                tmp = cur;
                block = Block.twoString;
                break;

            case block === Block.twoString && cur + next === '\\"':
                tmp += '"';
                i++;
                break;

            case block === Block.twoString && cur === '"':
                tmp += cur;
                finalize();
                block = Block.none;
                break;

            case block === Block.none && SYMBOLS[cur]:
                finalize();
                tmp = cur;
                finalize(true);
                break;

            case block === Block.none && EMPTY[cur]:
                finalize();
                break;

            default:
                tmp += cur;
        }

        if (block === Block.none && cur === '\n') {
            column = 1;
        } else {
            column++;
        }

        if (cur === '\n') {
            sameLine = false;
            curLine++;
        } else if (!EMPTY[cur]) {
            sameLine = true;
        }
    }

    return {
        tokens,
        lines,
        columns,
    };
}
