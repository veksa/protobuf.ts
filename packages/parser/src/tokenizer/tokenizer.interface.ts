export enum Block {
    none = 'none',
    comment = 'comment',
    multiComment = 'multiComment',
    singleQuoteString = 'singleQuoteString',
    doubleQuoteString = 'doubleQuoteString',
}
