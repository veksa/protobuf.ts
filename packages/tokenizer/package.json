{
  "name": "@protobuf.ts/tokenizer",
  "version": "0.0.1",
  "description": "Protocol Buffers Tokenizer for TypeScript.",
  "author": "Alex Khizhnyi <admin@devblog.pro>",
  "license": "BSD-3-Clause",
  "homepage": "https://github.com/veksa/protobuf.ts",
  "bugs": "https://github.com/veksa/protobuf.ts/issues",
  "main": "src/index.ts",
  "types": "src/index.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/veksa/protobuf.ts.git",
    "directory": "packages/tokenizer"
  },
  "keywords": [
    "protobuf",
    "protocol-buffers",
    "typescript",
    "tokenizer"
  ],
  "devDependencies": {
  },
  "dependencies": {
  }
}
