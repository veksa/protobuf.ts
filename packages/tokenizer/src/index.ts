export {tokenizer} from './tokenizer';
export {Symbol} from './tokenizer.constant';
