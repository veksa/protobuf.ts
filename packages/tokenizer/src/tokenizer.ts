import {Empty, Symbol} from './tokenizer.constant';
import {Block, IToken} from './tokenizer.interface';

export function tokenizer(source: string): IToken[] {
    const tokenList: IToken[] = [];

    let text = '';
    let block: Block = Block.none;
    let line = 1;
    let column = 1;
    let isSameLine = true;

    const finalize = (after = false, force = false) => {
        if (text.length > 0 || force) {
            tokenList.push({
                text,
                line,
                column: after
                    ? column
                    : column - text.length,
            });
            text = '';
        }
    };

    for (let i = 0; i < source.length; i++) {
        const cur = source[i];
        const next = source[i + 1];

        if (block === Block.none && cur + next === '//') {
            finalize();
            block = Block.comment;
            text += isSameLine ? '!//' : '//';
            finalize(true);
            i++;
            column++;
        } else if (block === Block.comment && cur === '\n') {
            block = Block.none;
            finalize(false, true);
        } else if (block === Block.none && cur + next === '/*') {
            block = Block.multiComment;
            finalize();
            text = '/*';
            finalize(true);
            i++;
            column++;
        } else if (block === Block.multiComment && cur + next === '*/') {
            block = Block.none;
            finalize(false, true);
            text = '*/';
            finalize(true);
            i++;
            column++;
        } else if (block === Block.none && cur === '\'') {
            finalize();
            text = cur;
            block = Block.singleQuoteString;
        } else if (block === Block.singleQuoteString && cur + next === '\\\'') {
            text += '\'';
            i++;
        } else if (block === Block.singleQuoteString && cur === '\'') {
            text += cur;
            finalize();
            block = Block.none;
        } else if (block === Block.none && cur === '"') {
            finalize();
            text = cur;
            block = Block.doubleQuoteString;
        } else if (block === Block.doubleQuoteString && cur + next === '\\"') {
            text += '"';
            i++;
        } else if (block === Block.doubleQuoteString && cur === '"') {
            text += cur;
            finalize();
            block = Block.none;
        } else if (block === Block.none && Symbol[cur]) {
            finalize();
            text = cur;
            finalize(true);
        } else if (block === Block.none && Empty[cur]) {
            finalize();
        } else {
            text += cur;
        }

        if (block === Block.none && cur === '\n') {
            column = 1;
        } else {
            column++;
        }

        if (cur === '\n') {
            isSameLine = false;
            line++;
        } else if (!Empty[cur]) {
            isSameLine = true;
        }
    }

    return tokenList;
}
