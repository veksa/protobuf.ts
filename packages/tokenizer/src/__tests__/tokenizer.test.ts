import {tokenizer} from '../tokenizer';

describe('tokenizer', () => {
    test('should ', () => {
        const actual = tokenizer('// test \n message');

        expect(actual).toMatchSnapshot();
    });
});
