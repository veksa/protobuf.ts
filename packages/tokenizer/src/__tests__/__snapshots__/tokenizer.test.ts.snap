// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenizer should  1`] = `
Object {
  "columns": Array [],
  "lines": Array [],
  "tokens": Array [],
}
`;
